lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -1149,21 +1124,38 @@ static void optimizeArithmeticNodes(Function *F) {
   // A worklist that contains the nodes to process.
   std::vector<Node *> worklist;
 
-  // Add all of the interesting nodes to the worklist.
-  for (auto &node : F->getNodes()) {
-    if (node.isArithmetic()) {
-      worklist.push_back(&node);
+  // Add all of the arithmetic nodes to the worklist, with a node's dependencies
+  // added after itself so they are processed before the node.
+  GraphPreOrderVisitor visitor(*F);
+  for (auto *N : visitor.getPreOrder()) {
```

Comment:
Nit: You could use `worklist.reserve(visitor.getPreOrder().size());` to avoid multiple dynamic `resize` calls when you add elements to the worklist.

Reply:
Good idea, thanks! Done.